Public/Src/Tools/SymbolDaemon/SymbolDaemon.cs

```diff
@@ -218,11 +229,117 @@ private static Client CreateClient(string serverMoniker, IClientConfig config)
+            {
+                var files = File.GetValues(conf.Config).ToArray();
+
+                // hashes are sent from BXL by serializing FileContentInfo's
+                var hashesWithLength = HashOptional.GetValues(conf.Config);                
+                var hashesOnly = hashesWithLength.Select(h => FileContentInfo.Parse(h).Hash).ToArray();
+                
+                var outputFile = SymbolMetadataFile.GetValue(conf.Config);
+                
+                IndexFilesAndStoreMetadataToFile(files, hashesOnly, outputFile);
+
+                return 0;
+            });
+
+        private static void IndexFilesAndStoreMetadataToFile(string[] files, ContentHash[] hashes, string outputFile)
+        {
+            Contract.Assert(files.Length == hashes.Length);
+            Contract.Assert(files.Length > 0);
+            Contract.Assert(!string.IsNullOrEmpty(outputFile));
+            Contract.Assert(!hashes.Any(hash => hash.HashType != HashType.Vso0), "Unsupported hash type");
+
+            var indexer = new SymbolIndexer(SymbolAppTraceSource.SingleInstance);
+            var symbolsMetadata = new Dictionary<ContentHash, HashSet<DebugEntryData>>();
+
+            for (int i = 0; i < files.Length; i++)
+            {
+                HashSet<DebugEntryData> symbols;
+                if (!symbolsMetadata.TryGetValue(hashes[i], out symbols))
+                {
+                    symbols = new HashSet<DebugEntryData>(DebugEntryDataComparer.Instance);
+                    symbolsMetadata.Add(hashes[i], symbols);
+                }               
+
+                // Index the file. It might not contain any symbol data. In this case, we will have an empty set.
+                symbols.UnionWith(indexer.GetDebugEntries(new System.IO.FileInfo(files[i]), calculateBlobId: false));               
+            }
+            
+            SerializeSymbolsMetadata(symbolsMetadata, outputFile);
+        }
+
+        /// <summary>
+        /// Serializes DebugEntryData's into a file.
+        /// </summary>        
+        public static void SerializeSymbolsMetadata(Dictionary<ContentHash, HashSet<DebugEntryData>> symbolsMetadata, string outputFile)
+        {
+            /*                
+                <number of hashes>
+                for each hash:
+                    <hash>
+                    <number of debug entries>
+                    [<debug entry>]
+             */
+             
+            using (var writer = new StreamWriter(outputFile))
+            {                
+                writer.WriteLine(symbolsMetadata.Count);
+
+                foreach (var kvp in symbolsMetadata)
+                {
+                    writer.WriteLine(kvp.Key.Serialize());
+                    writer.WriteLine(kvp.Value.Count);
+
+                    foreach (var debugEntry in kvp.Value)
+                    {
+                        writer.WriteLine(JsonConvert.SerializeObject(debugEntry));

```

Comment:
>JsonConvert.SerializeObject(debugEntry) [](start = 41, length = 39)

I'm not sure how complicated the DebugEntryData object is but I'd expect it to be more performant to couple this serialization to its current implementation rather than using a generic json serializer.